Realm/RLMSyncSubscription.mm

```diff
@@ -0,0 +1,483 @@
+                                         args:args];
+}
+
+- (void)addSubscriptionWithClassName:(NSString *)objectClassName
+                               where:(NSString *)predicateFormat
+                                args:(va_list)args {
+    [self addSubscriptionWithClassName:objectClassName
+                      subscriptionName:nil
+                             predicate:[NSPredicate predicateWithFormat:predicateFormat arguments:args]];
+}
+
+- (void)addSubscriptionWithClassName:(NSString *)objectClassName
+                    subscriptionName:(NSString *)name
+                               where:(NSString *)predicateFormat, ... {
+    va_list args;
+    va_start(args, predicateFormat);
+    return [self addSubscriptionWithClassName:objectClassName
+                             subscriptionName:name
+                                        where:predicateFormat
+                                         args:args];
+}
+
+- (void)addSubscriptionWithClassName:(NSString *)objectClassName
+                    subscriptionName:(NSString *)name
+                               where:(NSString *)predicateFormat
+                                args:(va_list)args {
+    [self addSubscriptionWithClassName:objectClassName
+                      subscriptionName:name
+                             predicate:[NSPredicate predicateWithFormat:predicateFormat arguments:args]];
+    
+}
+
+- (void)addSubscriptionWithClassName:(NSString *)objectClassName
+                           predicate:(NSPredicate *)predicate {
+    return [self addSubscriptionWithClassName:objectClassName
+                             subscriptionName:nil
+                                    predicate:predicate];
+}
+
+- (void)addSubscriptionWithClassName:(NSString *)objectClassName
+                    subscriptionName:(nullable NSString *)name
+                           predicate:(NSPredicate *)predicate {
+    return [self addSubscriptionWithClassName:objectClassName
+                             subscriptionName:name
+                                    predicate:predicate
+                               updateExisting:false];
+}
+
+- (void)addSubscriptionWithClassName:(NSString *)objectClassName
+                    subscriptionName:(nullable NSString *)name
+                           predicate:(NSPredicate *)predicate
+                      updateExisting:(BOOL)updateExisting {
+    [self verifyInWriteTransaction];
+    
+    RLMClassInfo& info = _realm->_info[objectClassName];
+    auto query = RLMPredicateToQuery(predicate, info.rlmObjectSchema, _realm.schema, _realm.group);
+    
+    if (name != nil) {
+        std::string str = std::string([name UTF8String]);
+        auto iterator = _mutableSubscriptionSet->find(str);
+        
+        if (iterator == _mutableSubscriptionSet->end() || updateExisting) {
+            _mutableSubscriptionSet->insert_or_assign(str, query);
+        }
+        else {
+            // Should this throw??
```

Comment:
This should either throw or return a value indicating whether or not it actually added a new subscription.